src/compute/src/server.rs

```diff
@@ -172,14 +219,154 @@ impl<'w, A: Allocate> Worker<'w, A> {
+        let idx = self.timely_worker.index();
+
+        {
+            let cmd_queue = Rc::clone(&cmd_queue);
+
+            self.timely_worker.dataflow::<u64, _, _>(move |scope| {
+                source(scope, "CmdSource", |capability, info| {
+                    // Send activator for this operator back
+                    let activator = scope.sync_activator_for(&info.address[..]);
+                    activator_tx.send(activator).expect("activator_tx working");
+
+                    //Hold onto capbility until we receive a disconnected error
+                    let mut cap_opt = Some(capability);
+
+                    move |output| {
+                        let mut disconnected = false;
+                        if let Some(cap) = cap_opt.as_mut() {
+                            let time = cap.time().clone();
+                            let mut session = output.session(&cap);
+
+                            loop {
+                                match command_rx.try_recv() {
+                                    Ok(cmd) => {
+                                        // Commands must never be sent to another worker. This
+                                        // implementation does not guarantee an ordering of events
+                                        // sent to different workers.
+                                        assert!(idx == 0);
+                                        session.give_iterator(
+                                            Self::split_command(cmd, peers).into_iter().enumerate(),
+                                        );
+                                    }
+                                    Err(TryRecvError::Disconnected) => {
+                                        disconnected = true;
+                                        break;
+                                    }
+                                    Err(TryRecvError::Empty) => {
+                                        break;
+                                    }
+                                };
+                            }
+                            cap.downgrade(&(time + 1));
+                        }
+
+                        if disconnected {
+                            cap_opt = None;
+                        }
+                    }
+                })
+                .unary_frontier::<Vec<()>, _, _, _>(
+                    Exchange::new(|(idx, _)| *idx as u64),
+                    "CmdReceiver",
+                    |_, _| {
+                        move |input, _| {
+                            let mut queue = cmd_queue.borrow_mut();
+                            if input.frontier().frontier().len() == 0 {
+                                queue.push_back(Err(TryRecvError::Disconnected))
+                            }
+                            while let Some((_, cmds)) = input.next() {
+                                for (_, cmd) in cmds.iter() {
+                                    queue.push_back(Ok(cmd.clone()));
```

Comment:
The usual pattern is to potentially avoid cloning by using [`RefOrMut::swap`](https://docs.rs/timely/latest/timely/communication/message/enum.RefOrMut.html#method.swap) to swap the input data into a local buffer and then `.drain(..)` that. Would that be possible here as well?

Reply:
Ah yes, thx